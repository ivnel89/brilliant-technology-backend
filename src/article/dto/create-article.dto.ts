import { ApiProperty } from "@nestjs/swagger";
import { IsString, IsUUID } from "class-validator";

export class CreateArticleDto {
    @ApiProperty()
    @IsString()
    content: string;
  
    @ApiProperty()
    @IsString()
    @IsUUID()
    authorId: string;
}
