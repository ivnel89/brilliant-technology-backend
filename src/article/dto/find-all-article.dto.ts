import { ApiProperty } from "@nestjs/swagger";
import { IsNumber, IsOptional, IsUUID, Max } from "class-validator";

export class FindAllArticleDto {
    @IsOptional()
    @IsNumber()
    @ApiProperty({
      required: false
    })
    offset?: number;

    @IsOptional()
    @IsNumber()
    @Max(250)
    @ApiProperty({
      required: false
    })
    limit?: number;

    @IsOptional()
    @ApiProperty({
      required: false
    })
    @IsUUID()
    requesterId?: string;
  }
  