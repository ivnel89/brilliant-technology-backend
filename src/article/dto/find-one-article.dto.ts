import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsUUID } from "class-validator";

export class FindOneArticleDto {
    @IsOptional()
    @ApiProperty({
      required: false
    })
    @IsUUID()
    requesterId?: string;
  }
  