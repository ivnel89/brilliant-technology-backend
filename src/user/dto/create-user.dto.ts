import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsString, IsUrl, Length } from "class-validator";

export class CreateUserDto {
  @ApiProperty()
  @IsString()
  @Length(1,256)
  firstName: string;

  @ApiProperty()
  @IsString()
  @Length(1,256)
  lastName: string;

  @ApiProperty()
  @IsUrl()
  @Length(1,2048)
  displayPicture: string;
}
