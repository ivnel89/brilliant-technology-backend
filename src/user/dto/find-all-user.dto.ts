import { ApiProperty } from "@nestjs/swagger";
import { IsNumber, IsOptional, Max } from "class-validator";

export class FindAllUserDto {
    @IsOptional()
    @IsNumber()
    @ApiProperty()
    offset?: number;

    @IsOptional()
    @IsNumber()
    @Max(250)
    @ApiProperty()
    limit?: number;
  }
  