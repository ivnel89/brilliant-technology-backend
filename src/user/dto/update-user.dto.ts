import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsString, IsUrl, Length } from "class-validator";

export class UpdateUserDto {
    @ApiProperty()
    @IsString()
    @IsOptional()
    @Length(1,256)
    firstName?: string;
  
    @ApiProperty({
        required: false
      })
    @IsString()
    @IsOptional()
    @Length(1,256)
    lastName?: string;
  
    @ApiProperty({
        required: false
      })
    @IsUrl()
    @IsOptional()
    @Length(1,2048)
    displayPicture?: string;

}
