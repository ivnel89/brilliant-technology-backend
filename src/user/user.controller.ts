import { Controller, Get, Post, Body, Patch, Param, Delete, Query } from '@nestjs/common';
import { UserService } from './user.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { FindAllUserDto } from './dto/find-all-user.dto';
import { UserContractMapper } from './mapper/user.mapper';
import { UserContract } from './contract/user.contract';

@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}
  private userContractMapper = new UserContractMapper();

  @Post()
  async create(@Body() createUserDto: CreateUserDto): Promise<UserContract> {
    const user = await this.userService
      .create(createUserDto);
    return this.userContractMapper.build(user);
  }

  @Get()
  async findAll(@Query() findAllUserDto?: FindAllUserDto): Promise<Array<UserContract>> {
    const users = await this.userService
      .findAll(findAllUserDto);
    return this.userContractMapper.buildArray(users);
  }

  @Get(':id')
  async findOne(@Param('id') id: string): Promise<UserContract> {
    const user = await this.userService
      .findOne(id);
    return this.userContractMapper.build(user);
  }

  @Patch(':id')
  async update(@Param('id') id: string, @Body() updateUserDto: UpdateUserDto): Promise<UserContract> {
    const user = await this.userService
      .update(id, updateUserDto);
    return this.userContractMapper.build(user);
  }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.userService.remove(id);
  }
}
