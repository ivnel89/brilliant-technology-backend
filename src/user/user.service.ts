import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import {  Repository, UpdateResult } from 'typeorm';
import { CreateUserDto } from './dto/create-user.dto';
import { FindAllUserDto } from './dto/find-all-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { User } from './entities/user.entity';

@Injectable()
export class UserService {
  constructor(
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) {}

  async create(createUserDto: CreateUserDto): Promise<User> {
    return this.userRepository.save(
      new User(
        createUserDto?.firstName,
        createUserDto?.lastName,
        createUserDto?.displayPicture,
      ),
    );
  }

  findAll(findAllUserDto: FindAllUserDto): Promise<Array<User>> {
    const { limit = 10, offset = 0} = findAllUserDto
    return this.userRepository.find({
      order: {
        lastModifiedDate: 'ASC',
      },
      skip: offset,
      take: limit,
    });
  }

  findOne(id: string): Promise<User> {
    return this.userRepository.findOne({
      where: {
        id,
      },
    });;
  }

  async update(id: string, updateUserDto: UpdateUserDto): Promise<User> {
    const user = await this.findOne(id);
    Object.keys(updateUserDto).forEach( key => {
      if(updateUserDto[key]){
        user[key] = updateUserDto[key];
      }
    })
    return this.userRepository.save(user);
  }

  remove(id: string): Promise<UpdateResult> {
    return this.userRepository.softDelete({ id });
  }
}
