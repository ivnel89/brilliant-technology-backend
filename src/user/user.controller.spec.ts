import { Test, TestingModule } from '@nestjs/testing';
import { plainToInstance } from 'class-transformer';
import { validate } from 'class-validator';
import { CreateUserDto } from './dto/create-user.dto';
import { FindAllUserDto } from './dto/find-all-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { createUserDto, findAllUserDto, mockUserId, updateUserDto } from './test/user.dto.mock';
import { UserRepository } from './test/user.repository.mock';
import { UserController } from './user.controller';
import { UserService } from './user.service';

describe('UserController', () => {
  let controller: UserController;
  let service: UserService;
  let serviceCreateSpy,
    serviceFindAllSpy,
    serviceFindOneSpy,
    serviceUpdateSpy,
    serviceRemoveSpy;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [UserController],
      providers: [
        UserService,
        UserRepository
      ],
    }).compile();

    controller = module.get<UserController>(UserController);
    service = module.get<UserService>(UserService);
    serviceCreateSpy = jest.spyOn(service, 'create');
    serviceFindAllSpy = jest.spyOn(service, 'findAll'); 
    serviceFindOneSpy = jest.spyOn(service, 'findOne'); 
    serviceUpdateSpy = jest.spyOn(service, 'update'); 
    serviceRemoveSpy = jest.spyOn(service, 'remove');
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  describe('create', () => {
    const serviceCreateError = "serviceCreateError";
    it('should call create from userService with right params', async () => {
      await expect(controller.create(createUserDto)).resolves.toMatchSnapshot().then(() => {
        expect(serviceCreateSpy).toBeCalledWith(createUserDto);
      })
    })
    it('should validate createUserDto', async () => { 
      const invalidInput = new CreateUserDto();
      invalidInput.displayPicture = 'notAnUrl';
      invalidInput.firstName = 999 as any;
      const errors = await validate(plainToInstance(CreateUserDto,invalidInput))
      expect(errors.length).toBe(3)
    })
    it('should throw an error when create from userService throws an error', async () => {
      serviceCreateSpy.mockRejectedValueOnce(serviceCreateError);
      await expect(controller.create(createUserDto)).rejects.toBe(
        serviceCreateError,
      );
    })
  })

  describe('findAll', () => {
    const serviceFindAllError = "serviceFindAllError";
    it('should call findAll from userService with right params', async () => {
      await expect(controller.findAll(findAllUserDto)).resolves.toMatchSnapshot().then(() => {
        expect(serviceFindAllSpy).toBeCalledWith(findAllUserDto);
      })
    })
    it('should validate findAllUserDto', async () => { 
      const invalidInput = new FindAllUserDto();
      invalidInput.limit = NaN;
      invalidInput.offset = "twentyTwo" as any;
      const errors = await validate(plainToInstance(FindAllUserDto,invalidInput))
      expect(errors.length).toBe(2)
    })
    it('should throw an error when findAll from userService throws an error', async () => {
      serviceFindAllSpy.mockRejectedValueOnce(serviceFindAllError);
      await expect(controller.findAll(findAllUserDto)).rejects.toBe(
        serviceFindAllError,
      );
    })
  })

  describe('findOne', () => {
    const serviceFindOneError = "serviceFindOneError";
    it('should call findOne from userService with right params', async () => {
      await expect(controller.findOne(mockUserId)).resolves.toMatchSnapshot().then(() => {
        expect(serviceFindOneSpy).toBeCalledWith(mockUserId);
      })
    })
    it('should throw an error when findOne from userService throws an error', async () => {
      serviceFindOneSpy.mockRejectedValueOnce(serviceFindOneError);
      await expect(controller.findOne(mockUserId)).rejects.toBe(
        serviceFindOneError,
      );
    })
  })

  describe('update', () => {
    const serviceUpdateError = "serviceUpdateError";
    it('should call update from userService with right params', async () => {
      await expect(controller.update(mockUserId,updateUserDto)).resolves.toMatchSnapshot().then(() => {
        expect(serviceUpdateSpy).toBeCalledWith(mockUserId,updateUserDto);
      })
    })
    it('should validate updateUserDto', async () => { 
      const invalidInput = new UpdateUserDto();
      invalidInput.displayPicture = 'notAnUrl';
      invalidInput.firstName = 999 as any;
      const errors = await validate(plainToInstance(UpdateUserDto,invalidInput))
      expect(errors.length).toBe(2)
    })
    it('should throw an error when create from userService throws an error', async () => {
      serviceUpdateSpy.mockRejectedValueOnce(serviceUpdateError);
      await expect(controller.update(mockUserId, updateUserDto)).rejects.toBe(
        serviceUpdateError,
      );
    })
  })

  describe('remove', () => {
    const serviceRemoveError = "serviceRemoveError";
    it('should call remove from userService with right params', async () => {
      await expect(controller.remove(mockUserId)).resolves.toMatchSnapshot().then(() => {
        expect(serviceRemoveSpy).toBeCalledWith(mockUserId);
      })
    })
    it('should throw an error when remove from userService throws an error', async () => {
      serviceRemoveSpy.mockRejectedValueOnce(serviceRemoveError);
      await expect(controller.remove(mockUserId)).rejects.toBe(
        serviceRemoveError,
      );
    })
  })

});
