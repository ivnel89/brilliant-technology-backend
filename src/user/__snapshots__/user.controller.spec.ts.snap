// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController create should call create from userService with right params 1`] = `
UserContract {
  "displayPicture": "www.example.com/image.png",
  "firstName": "Foo",
  "id": undefined,
  "lastName": "Bar",
}
`;

exports[`UserController findAll should call findAll from userService with right params 1`] = `
Array [
  UserContract {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "id": undefined,
    "lastName": "Bar",
  },
]
`;

exports[`UserController findOne should call findOne from userService with right params 1`] = `
UserContract {
  "displayPicture": "www.example.com/img.png",
  "firstName": "Foo",
  "id": undefined,
  "lastName": "Bar",
}
`;

exports[`UserController remove should call remove from userService with right params 1`] = `"Deleted"`;

exports[`UserController update should call update from userService with right params 1`] = `
UserContract {
  "displayPicture": "www.example.com/img.png",
  "firstName": "Foo",
  "id": undefined,
  "lastName": "Far",
}
`;
